pkg/ingester/user_tsdb.go

```diff
@@ -467,3 +499,99 @@ func (u *userTSDB) releaseAppendLock(acquireState tsdbState) {
 		u.inFlightAppendsStartedBeforeForcedCompaction.Done()
 	}
 }
+
+func (u *userTSDB) OwnedSeriesAndShards() (int, int) {
+	u.ownedSeriesMtx.Lock()
+	defer u.ownedSeriesMtx.Unlock()
+
+	return u.ownedSeriesCount, u.ownedSeriesShardSize
+}
+
+func (u *userTSDB) TriggerRecomputeOwnedSeries(reason string) {
+	u.requiresOwnedSeriesUpdate.CompareAndSwap("", reason)
+}
+
+func (u *userTSDB) RecomputeOwnedSeries(shardSize int, reason string, logger log.Logger) {
+	// We need to recompute owned series, ie. how many series in this user's Head are owned by this ingester (according to
+	// current token ranges), and updates both ownedSeries and ownedSeriesShardSize.
+
+	start := time.Now()
+	repeats := 0
+
+	ownedPrev, shardSizePrev := u.OwnedSeriesAndShards()
+
+	var ownedNew int
+	for {
+		ownedNew = u.computeOwnedSeries()
+
+		// Try to update the values, but only if no new series were added in the meantime.
+		u.ownedSeriesMtx.Lock()
+		if u.ownedSeriesCount == ownedPrev {
```

Comment:
I'm not super convinced of repeating even when the diff is small. What if we tolerate a small drift, instead?

Reply:
What is "small drift" in your opinion? I am thinking less than 100 new series.

Reply:
Agreed on 1000 on Slack